src/test/java/com/bookstore/controller/BookControllerTest.java

```diff
@@ -0,0 +1,179 @@
+
+        // When
+        MvcResult result = mockMvc.perform(post("/books")
+                        .content(jsonRequest)
+                        .contentType(MediaType.APPLICATION_JSON_VALUE))
+                .andExpect(status().isOk())
+                .andReturn();
+        BookDto actual = objectMapper.readValue(result.getResponse().getContentAsString(),
+                BookDto.class);
+
+        // Then
+        assertTrue(EqualsBuilder.reflectionEquals(expected, actual,
+                "categoryIds", "coverImage", "id", "description"));
+    }
+
+    @Test
+    @DisplayName("""
+            Given a valid role
+            When a GET request is sent to /api/books/
+            Then return a list of Book Dtos
+            """)
+    @WithMockUser(username = "username")
+    @Sql(scripts = {"classpath:/database/books/add-one-category.sql",
+            "classpath:/database/books/add-two-books-with-categories.sql"},
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    public void findAll_validRole_returnListOfBookDtos() throws Exception {
+        // Given
+        List<BookDto> expected = bootstrapBookDtoList();
+
+        // When
+        MvcResult result = mockMvc.perform(get("/books")
+                        .contentType(MediaType.APPLICATION_JSON_VALUE))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        // Then
+        List<BookDto> actual = objectMapper.readValue(result.getResponse().getContentAsString(),
+                new TypeReference<>() {
+                });
+        assertEquals(expected.size(), actual.size());
+        for (int i = 0; i < actual.size(); i++) {
+            BookDto actualBook = actual.get(i);
+            BookDto expectedBook = expected.get(i);
+            assertTrue(EqualsBuilder.reflectionEquals(expectedBook, actualBook, "categoryIds",
+                    "coverImage",
+                    "id",
+                    "description",
+                    "price"));
+            assertTrue(compareTruncatedRoundedDecimals(expectedBook.getPrice(),
+                    actualBook.getPrice()));
+            assertEquals(expectedBook.getCategoryIds().size(), actualBook.getCategoryIds().size());
+        }
+    }
+
+    @WithMockUser(username = "user")
+    @Test
+    @DisplayName("""
+            Given a negative id
+            When a GET request is sent to /api/books/{id}
+            Then throw an exception
+            """)
+    void getBookById_negativeId_throwException() throws Exception {
+        // Given
+        Long invalidId = -1L;
+
+        // When
+        mockMvc.perform(get("/books/{id}", invalidId)
+                        .contentType(MediaType.APPLICATION_JSON_VALUE))
+                .andExpect(status().isNotFound())
+                .andExpect(result -> assertInstanceOf(EntityNotFoundException.class,
+                        result.getResolvedException()));
+
+        // Then
```

Comment:
remove Then if you don’t  have specific logic for this block (check all test cases)